python/lsst/daf/butler/core/dimensions/dataId.py

```diff
@@ -0,0 +1,351 @@
+            self._entries = {element: {} for element in self._dimensions.elements}
+
+        # Return the new instance, invoking __init__ to do further updates.
+        return self
+
+    def __init__(self, dataId=None, *, dimensions=None, dimension=None, universe=None, region=None,
+                 entries=None, extra=None, **kwds):
+        if dataId is None:
+            dataId = {}
+
+        if entries is not None:
+            for element, subdict in entries.items():
+                self.entries[element].update(subdict)
+
+        missing = self.dimensions.links - self._links.keys()
+        for linkName in missing:
+            # Didn't get enough key-value pairs to identify all dimensions
+            # from the links; look in
+            # entries for those.
+            for element in self.dimensions.withLink(linkName):
+                try:
+                    self._links[linkName] = self.entries[element][linkName]
+                    break
+                except KeyError:
+                    pass
+            else:
+                raise LookupError(f"No value found for link '{linkName}'")
+
+        # If we got an explicit region argument, use it.
+        if region is not None:
+            self.region = region
+            self.entries[self.dimensions.getRegionHolder()]["region"] = region
+
+        # Entries should contain link fields as well, so transfer them from
+        # 'extra' + 'kwds'.  Also transfer from 'links' iff it's not a DataId;
+        # if it is, we can safely assume the transfer has already been done.
+
+        def addLinksToEntries(items):
+            for linkName, linkValue in items:
+                try:
+                    associated = self.dimensions.withLink(linkName)
+                except KeyError:
+                    # This isn't a link. If an explicit dimension was
+                    # provided, assume these fields are metadata for that
+                    # dimension.
+                    if dimension is not None:
+                        self.entries[self.dimension][linkName] = linkValue
+                    else:
+                        raise
+                for element in associated:
+                    if element in self.dimensions:
+                        self.entries[element][linkName] = linkValue
+
+        if extra is not None:
+            addLinksToEntries(extra.items())
+        addLinksToEntries(kwds.items())
+        if not isinstance(dataId, DataId):
+            addLinksToEntries(dataId.items())
+
+        # If we still haven't got a region, look for one in entries.
+        if self.region is None:
+            holder = self.dimensions.getRegionHolder()
+            if holder is not None:
+                self.region = self.entries[holder].get("region", None)
+
+    @property
+    def dimensions(self):
+        """The dimensions this `DataId` identifies (`DimensionGraph`).
+        """
+        return self._dimensions
+
+    @property
+    def dimension(self):
```

Comment:
Thinking about it a bit more, I think the concept of having a leaf dimension probably belongs on `DimensionGraph` itself, and it may make sense to generalize to multiple leaf dimensions.  I'll give that a try.

Reply:
I've added a `.leaves` property to `DimensionGraph`, and that (slightly) simplifies the `DataId` and probably helps with some `SqlRegistry` code, too.